infrastructure/smart-contracts/test/factory/AchievementsFactoryTests.mjs

```diff
@@ -359,10 +427,86 @@ export function AchievementsFactoryTests(deployInfrastructure) {
+
+            //assert contract state and events
+            expect(await achievementsFactory2.test()).to.equal(testNum);
+            expect(contractProducedEventLog.fragment.name).to.equal("ContractProduced");
+            expect(contractProducedEventLog.args[1]).to.equal(gameId);
+            expect(contractProducedEventLog.args[2]).to.equal(await addr1.getAddress());
+            expect(baseURI).to.equal(await tokenContract.uri(0));
+            expect(await tokenContract.factoryAddress()).to.equal(await achievementsFactory.getAddress());
+            expect(await achievementsFactory.productionCount()).to.equal(1);
+            expect(await achievementsFactory.contractsById(gameId)).to.equal(await tokenContract.getAddress());
         });
 
+        it("should upgrade all Achievements contracts to new version", async function() {
+            const {
+                achievementsFactory,
+                achievementsBeacon,
+                addr1,
+                addr2
+            } = await loadFixture(deployInfrastructure);
+
+            //produce new v1 token contract
+            const gameId = "test-game-id";
+            const produceContractTrx = await achievementsFactory.connect(addr1).produceContract(gameId, baseURI);
+            const produceContractRec = await produceContractTrx.wait();
+            const tokenContractAddress = produceContractRec.logs[2].args[0];
+            const AchievementsContractFactory = await ethers.getContractFactory("Achievements");
+            const tokenContractV1 = await AchievementsContractFactory.attach(tokenContractAddress);
+
+            //mint tokens on v1 contract
+            const toAddress = await addr1.getAddress();
+            const tokenId = 0;
+            await tokenContractV1.connect(addr1).mint(toAddress, tokenId);
+
+            //deploy token contract v2 implementation
+            const Achievements2Impl = await ethers.deployContract("Achievements2");
+            await Achievements2Impl.waitForDeployment();
+            const achievements2ImplAddress = await Achievements2Impl.getAddress();
+
+            //validate that v2 is upgrade safe
+            const Achievements2ContractFactory = await ethers.getContractFactory("Achievements2");
+            const achievementsImplReference = await upgrades.forceImport(tokenContractAddress, AchievementsContractFactory);
+            await upgrades.validateUpgrade(achievementsImplReference, Achievements2ContractFactory);
+
+            //upgrade the implementation address
+            await achievementsBeacon.update(achievements2ImplAddress);
+            const impl2 = await achievementsBeacon.implementation();
+            expect(impl2).to.equal(achievements2ImplAddress);
+
+            //initialize v2 token contract
+            const testNum = 55n;
+            const tokenContractV2 = await Achievements2ContractFactory.attach(tokenContractAddress);
+            await tokenContractV2.initialize(testNum);
```

Comment:
We should have a new function `testcaseUpgrade() pure` that will only exist in the test upgrade implementation.

Then you can call this on V1 expect it to revert and call it again after the upgrade and make sure it returns a valid response.